testing/dsp_testing/test_beamforming.py

```diff
@@ -1,123 +1,693 @@
+                    if filetype == tx_filetype:  # tx data has sequence number 0 for all 
+                        decimated_data = record_dict['data'][0,:,dm_start_sample:dm_end_sample:dm_rate]
+                    else:
+                        decimated_data = record_dict['data'][sequence_num,:,dm_start_sample:dm_end_sample:dm_rate]
+                else:
+                    raise Exception('Not sure how to decimate with the dimensions of this data: {}'.format(record_dict['data_descriptors']))
+                
+            else:
+                if data_description_list == ['num_antenna_arrays', 'num_sequences', 'num_beams', 'num_samps']:
+                    decimated_data = record_dict['data'][0,sequence_num,:,:] # only main array
+                    intf_decimated_data = record_dict['data'][1,sequence_num,:,:]
+                elif data_description_list == ['num_antennas', 'num_sequences', 'num_samps']:
+                    decimated_data = record_dict['data'][:,sequence_num,:] # first sequence only, all antennas.
+                elif data_description_list == ['num_sequences', 'num_antennas', 'num_samps']:
+                    if filetype == tx_filetype:
+                        decimated_data = record_dict['data'][0,:,:] # first sequence only, all antennas.
+                    else:
+                        decimated_data = record_dict['data'][sequence_num,:,:] # first sequence only, all antennas.
+                else:
+                    raise Exception('Unexpected data dimensions: {}'.format(record_dict['data_descriptors']))
+
+            # STEP 2: BEAMFORM ANY UNBEAMFORMED DATA
+            if filetype != bfiq_filetype:
+                # need to beamform the data. 
+                antenna_list = []
+                #print(decimated_data.shape)
+                if data_description_list == ['num_antennas', 'num_sequences', 'num_samps']:
+                    for antenna in range(0, record_dict['data'].shape[0]):
+                        antenna_list.append(decimated_data[antenna,:])
+                    antenna_list = np.array(antenna_list)
+                elif data_description_list == ['num_sequences', 'num_antennas', 'num_samps']:
+                    for antenna in range(0, record_dict['data'].shape[1]):
+                        antenna_list.append(decimated_data[antenna,:])
+                    antenna_list = np.array(antenna_list)
+                else:
+                    raise Exception('Not sure how to beamform with the dimensions of this data: {}'.format(record_dict['data_descriptors']))
+
+                # beamform main array antennas only. 
+                main_antennas_mask = (record_dict['antennas_present'] < main_antenna_count)
+                intf_antennas_mask = (record_dict['antennas_present'] >= main_antenna_count)
+                decimated_beamformed_data = beamform(antenna_list[main_antennas_mask][:].copy(), beam_azms, freq) 
```

Comment:
I think the [:] already makes a copy of the list, so you wouldn't have to call .copy() on it